asgiref/current_thread_executor.py

```
@@ -68,20 +70,41 @@ def run_until_future(self, future: "Future[Any]") -> None:
             raise RuntimeError(
                 "You cannot run CurrentThreadExecutor from a different thread"
             )
-        future.add_done_callback(self._work_queue.put)
-        # Keep getting and running work items until we get the future we're waiting for
-        # back via the future's done callback.
-        try:
-            while True:
+
+        is_done = False  # synchronized by _work_ready
+
+        def done(future: "Future[Any]") -> None:
+            with self._work_ready:
+                nonlocal is_done
+                is_done = True
+                self._work_ready.notify()
+
+        future.add_done_callback(done)
+        # Keep getting and running work items until the future we're waiting for
+        # is done.
+        while True:
+            with self._work_ready:
+                while not is_done and not self._work_items:
+                    self._work_ready.wait()
+                if is_done:
+                    self._broken = True
+                    break
                 # Get a work item and run it
-                work_item = self._work_queue.get()
-                if work_item is future:
-                    return
-                assert isinstance(work_item, _WorkItem)
-                work_item.run()
-                del work_item
-        finally:
-            self._broken = True
+                work_item = self._work_items.popleft()
+            work_item.run()
+            del work_item
+
+        # Resubmit remaining work to the closest running CurrentThreadExecutor
+        # on the stack, if any.
+        if self._work_items:
+            executor = self._old_executor
+            while executor is not None:
+                with executor._work_ready:
+                    if not executor._broken:
+                        executor._work_items.extend(self._work_items)
+                        executor._work_ready.notify()
+                        break
+                executor = executor._old_executor
```

Alternatively, we could execute the remaining work immediately. This would block the `async_to_sync` caller for longer, but would avoid the possibility of silently discarding the work if we don’t find a running `CurrentThreadExecutor`:

```py
        while self._work_items:
            work_item = self._work_items.popleft()
            work_item.run()
```

Which strategy makes more sense?

---

You have a correct explanation there for why `submit()` needs to migrate newly-submitted jobs up the stack; we’d still add that behavior either way. However, the question here is whether `run_until_future()` needs to further migrate its own already-submitted jobs up the stack, or whether it should simply execute them on the spot.

In other words: if the user runs
`async_to_sync` [A] → `sync_to_async` → `async_to_sync` [B] → `create_task` → `sync_to_async` → job [J],
and [J] hasn’t run yet, should we execute [J] as [B] is returning, or migrate it up to [A]?

Similarly, if the user runs
`async_to_sync` [A] → `create_task` → `sync_to_async` → job [J],
and [J] hasn’t run yet, should we execute [J] as [A] is returning, or discard it completely?

---

@andersk This block (from ln100) is untested. Could you add an example showing what's at stake here? 

> ...the possibility of silently discarding the work... 

So _errors shouldn't pass silently_ (or something) — if I arrive here is it not an error I want to know about? 🤔

---

This block is tested, by the new tests `tests/test_sync.py::test_nested_task` (from #492) and `tests/test_sync.py::test_double_nested_task`.

It seems like asgiref has never considered discarded work to be an error before—but yeah, I think I agree we should never discard work. After thinking about it some more, I came up with another test case where discarded work may lead to a deadlock. So I’ll switch the strategy as proposed and add that test case.